hydradx/tests/test_omnipool_amm.py

```diff
@@ -143,14 +148,57 @@ def test_add_risk_liquidity(old_state):
         assert oamm.price_i(old_state, j) == pytest.approx(oamm.price_i(new_state, j))
     assert old_state['R'][i] / old_state['S'][i] == pytest.approx(new_state['R'][i] / new_state['S'][i])
 
+    assert old_state['L'] / sum(old_state['Q']) == pytest.approx(new_state['L'] / sum(new_state['Q']))
+
+    # check enforcement of agent's spending limit
+    new_state, new_agents = oamm.add_risk_liquidity(old_state, old_agents, LP_id, old_agents[LP_id]['r'][0] + 1, 0)
+    assert new_state, new_agents == (old_state, old_agents)
+    new_state, new_agents = oamm.add_risk_liquidity(old_state, old_agents, LP_id, old_agents[LP_id]['r'][0] - 1, 0)
+    assert new_state, new_agents != (old_state, old_agents)
+
+    # check enforcement of overall TVL cap
+    stablecoin_index = old_state['stablecoin_index']
+    TVL = sum([
+        old_state['Q'][i] * old_state['R'][stablecoin_index] / old_state['Q'][stablecoin_index]
+        for i in range(len(old_state['token_list']))
+    ])
+    assert TVL == sum(old_state['T'])
+    old_state['C'] = TVL
+    new_state, new_agents = oamm.add_risk_liquidity(old_state, old_agents, LP_id, 1, 0)
+    assert new_state, new_agents == (old_state, old_agents)
+
+    # check enforcement of per-asset weight limit
+    total_Q = sum(old_state['Q'])
+    for i in range(nof_tokens):
+        old_state['Q'][i] = total_Q / nof_tokens
+        old_state['R'][i] = total_Q / nof_tokens
+        old_state['T'][i] = total_Q / nof_tokens
+    i = 0
+    asset_price = old_state['R'][i] / old_state['Q'][i]
+    max_amount = (old_state['O'][i] - 1 / nof_tokens) / old_state['O'][i] * total_Q * asset_price
+    # make sure checks other than weight limit will pass
+    old_state['C'] = total_Q * 2
+    old_agents[LP_id]['r'][i] = max_amount * 2
+
+    new_state, new_agents = oamm.add_risk_liquidity(old_state, old_agents, LP_id, max_amount + 1, i)
+    if not (new_state, new_agents) == (old_state, old_agents):
+        raise ValueError(f'illegal transaction passed against weight limit in {i}')
```

Comment:
Right, I put that 'raise' in there because (maybe due to my test configuration) it was just failing silently leaving me to guess what went wrong. But I think I have fixed that at this point.

Reply:
Ok, I changed it